{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['https://cellxgene.cziscience.com/d/cellular_census_of_human_lungs_alveoli_and_parenchyma-15.cxg/', 'https://cellxgene.cziscience.com/d/cellular_census_of_human_lungs_bronchi-17.cxg/', 'https://cellxgene.cziscience.com/d/cellular_census_of_human_lungs_nasal-16.cxg/', 'https://www.covid19cellatlas.org/deprez19/', 'https://cellxgene.cziscience.com/d/Single_cell_gene_expression_profiling_of_SARS_CoV_2_infected_human_cell_lines_H1299-27.cxg/', 'https://cellxgene.cziscience.com/d/Single_cell_gene_expression_profiling_of_SARS_CoV_2_infected_human_cell_lines_Calu_3-28.cxg/', 'https://www.covid19cellatlas.org/lukassen20-lung/', 'https://www.covid19cellatlas.org/lukassen20-hbec/', 'https://www.covid19cellatlas.org/miller20/', 'https://cellxgene.cziscience.com/d/ischaemic_sensitivity_of_human_tissue_by_single_cell_RNA_seq_lung-18.cxg/', 'https://www.covid19cellatlas.org/baron16/', 'https://www.covid19cellatlas.org/cheng18/', 'https://www.covid19cellatlas.org/guo18/', 'https://www.covid19cellatlas.org/habib17/', 'https://www.covid19cellatlas.org/aldinger20/', 'https://www.covid19cellatlas.org/henry18/', 'https://cellxgene.cziscience.com/d/gut_cell_atlas-24.cxg/', 'https://www.covid19cellatlas.org/lako/', 'https://www.covid19cellatlas.org/litvinukova20/', 'https://www.covid19cellatlas.org/macparland18/', 'https://cellxgene.cziscience.com/d/human_fetal_liver_single_cell_transcriptome-13.cxg/', 'https://www.covid19cellatlas.org/miao20/', 'https://www.covid19cellatlas.org/byrd20/', 'https://www.covid19cellatlas.org/warner20/', 'https://www.covid19cellatlas.org/smillie19/', 'https://cellxgene.cziscience.com/d/spatio_temporal_immune_zonation_of_the_human_kidney-21.cxg/', 'https://www.covid19cellatlas.org/voigt19/', 'https://www.covid19cellatlas.org/menon19/', 'https://www.covid19cellatlas.org/lukowski19/', 'https://cellxgene.cziscience.com/d/fetal_maternal_interface_10x-22.cxg/', 'https://cellxgene.cziscience.com/d/fetal_maternal_interface_smartseq2-23.cxg/', 'https://www.covid19cellatlas.org/wang20-colon/', 'https://www.covid19cellatlas.org/martin19/', 'https://www.covid19cellatlas.org/wang20-ileum/', 'https://www.covid19cellatlas.org/wang20-rectum/', 'https://cellxgene.cziscience.com/d/cell_atlas_of_thymic_development-14.cxg/', 'https://cellxgene.cziscience.com/d/ischaemic_sensitivity_of_human_tissue_by_single_cell_RNA_seq_spleen-19.cxg/', 'https://cellxgene.cziscience.com/d/ischaemic_sensitivity_of_human_tissue_by_single_cell_RNA_seq_oesophagus-20.cxg/']\n"
     ]
    }
   ],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "from urllib.request import Request, urlopen\n",
    "import re\n",
    "\n",
    "req = Request(\"https://www.covid19cellatlas.org/index.healthy.html\")\n",
    "html_page = urlopen(req)\n",
    "\n",
    "soup = BeautifulSoup(html_page, \"lxml\")\n",
    "\n",
    "links = []\n",
    "for link in soup.findAll('a'):\n",
    "    try:          \n",
    "        classes = link.get(\"class\")\n",
    "        if 'btn-primary' in classes:\n",
    "            if \".cxg\" in link.get(\"href\"):\n",
    "                links.append(link.get(\"href\"))\n",
    "            else:\n",
    "                links.append(\"https://www.covid19cellatlas.org\" + link.get(\"href\"))\n",
    "    except:\n",
    "        continue\n",
    "        \n",
    "print(links)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://cellxgene.cziscience.com/d/cellular_census_of_human_lungs_alveoli_and_parenchyma-15.cxg/'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "links[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running for https://cellxgene.cziscience.com/d/cellular_census_of_human_lungs_alveoli_and_parenchyma-15.cxg/\n"
     ]
    },
    {
     "ename": "WebDriverException",
     "evalue": "Message: TypeError: this.curBrowser.contentBrowser is null\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mWebDriverException\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-81830ca88161>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     56\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0murl\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mlinks\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     57\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"Running for {url}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 58\u001b[0;31m     \u001b[0mcell_types\u001b[0m \u001b[0;34m=\u001b[0m  \u001b[0mget_cell_types\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     59\u001b[0m     \u001b[0mall_datasets\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcell_types\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-3-81830ca88161>\u001b[0m in \u001b[0;36mget_cell_types\u001b[0;34m(url)\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m     \u001b[0mdriver\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mwebdriver\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mFirefox\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 15\u001b[0;31m     \u001b[0mdriver\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m     \u001b[0mall_cell_types\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/selenium/webdriver/remote/webdriver.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, url)\u001b[0m\n\u001b[1;32m    331\u001b[0m         \u001b[0mLoads\u001b[0m \u001b[0ma\u001b[0m \u001b[0mweb\u001b[0m \u001b[0mpage\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mcurrent\u001b[0m \u001b[0mbrowser\u001b[0m \u001b[0msession\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    332\u001b[0m         \"\"\"\n\u001b[0;32m--> 333\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mCommand\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGET\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m'url'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    334\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    335\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/selenium/webdriver/remote/webdriver.py\u001b[0m in \u001b[0;36mexecute\u001b[0;34m(self, driver_command, params)\u001b[0m\n\u001b[1;32m    319\u001b[0m         \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcommand_executor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdriver_command\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    320\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 321\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merror_handler\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcheck_response\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresponse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    322\u001b[0m             response['value'] = self._unwrap_value(\n\u001b[1;32m    323\u001b[0m                 response.get('value', None))\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/selenium/webdriver/remote/errorhandler.py\u001b[0m in \u001b[0;36mcheck_response\u001b[0;34m(self, response)\u001b[0m\n\u001b[1;32m    240\u001b[0m                 \u001b[0malert_text\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'alert'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'text'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    241\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mexception_class\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmessage\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mscreen\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstacktrace\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0malert_text\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 242\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mexception_class\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmessage\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mscreen\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstacktrace\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    243\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    244\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_value_or_default\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdefault\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mWebDriverException\u001b[0m: Message: TypeError: this.curBrowser.contentBrowser is null\n"
     ]
    }
   ],
   "source": [
    "from selenium import webdriver\n",
    "from selenium.webdriver.support.ui import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.common.exceptions import TimeoutException\n",
    "import time\n",
    "\n",
    "\n",
    "def get_cell_types(url):\n",
    "    \"\"\"\n",
    "    Scrapes cell types from COVID-19 visualization portal\n",
    "    \"\"\"\n",
    "\n",
    "    driver = webdriver.Firefox()\n",
    "    driver.get(url)\n",
    "\n",
    "    all_cell_types = []\n",
    "    try:\n",
    "        print(\"Waiting 20 seconds\")\n",
    "        time.sleep(20)\n",
    "        \n",
    "        cookies=\"button.bp3-intent-primary:nth-child(1) > span:nth-child(1)\"\n",
    "        driver.find_element_by_css_selector(cookies).click()\n",
    "        \n",
    "        \n",
    "        cell_type_names = [\"annotation:category-expand\", \n",
    "                           \"broad_celltype:category-expand\", \n",
    "                           \"BroadCellType:category-expand\", \n",
    "                           \"CellType:category-expand\", \n",
    "                           \"celltype:category-expand\", \n",
    "                           \"cell_type:category-expand\"]\n",
    "        \n",
    "        for i in driver.find_elements_by_xpath(\"//span[@data-testid]\"):\n",
    "            title = i.get_attribute(\"data-testid\")\n",
    "            print(title)\n",
    "\n",
    "            if title in cell_type_names:\n",
    "                try:\n",
    "                    i.click()\n",
    "                except:\n",
    "                    print(\"oops\")\n",
    "\n",
    "        for i in driver.find_elements_by_xpath(\"//span[@aria-label]\"):\n",
    "            label = i.get_attribute(\"aria-label\")\n",
    "            all_cell_types.append(label)\n",
    "        driver.quit()\n",
    "        print(all_cell_types)\n",
    "        return(all_cell_types)\n",
    "\n",
    "    except:\n",
    "        return(\"failed\")\n",
    "    \n",
    "    \n",
    "\n",
    "all_datasets = {}\n",
    "for url in links:\n",
    "    print(f\"Running for {url}\")\n",
    "    cell_types =  get_cell_types(url)\n",
    "    all_datasets[url] = set(cell_types)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://cellxgene.cziscience.com/d/cellular_census_of_human_lungs_nasal-16.cxg/'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Waiting 10 seconds\n"
     ]
    }
   ],
   "source": [
    "driver = webdriver.Firefox()\n",
    "driver.get(url)\n",
    "\n",
    "all_cell_types = []\n",
    "print(\"Waiting 10 seconds\")\n",
    "time.sleep(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "header\n",
      "BroadCellType:category-expand\n",
      "BroadCellType:category-label\n",
      "CellType:category-expand\n",
      "CellType:category-label\n",
      "Donor:category-expand\n",
      "Donor:category-label\n",
      "Sample:category-expand\n",
      "Sample:category-label\n",
      "cellset-count-1\n",
      "cellset-count-2\n"
     ]
    }
   ],
   "source": [
    "#cookies=\"button.bp3-intent-primary:nth-child(1) > span:nth-child(1)\"\n",
    "#driver.find_element_by_css_selector(cookies).click()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'https://cellxgene.cziscience.com/d/cellular_census_of_human_lungs_alveoli_and_parenchyma-15.cxg/': {'A cellular census of human lungs identifies novel cell states in health and in asthma - parenchyma',\n",
       "  'B cells',\n",
       "  'Basal',\n",
       "  'Basal 1',\n",
       "  'Basal 2',\n",
       "  'BroadCellType',\n",
       "  'CellType',\n",
       "  'Ciliated',\n",
       "  'Ciliated 1',\n",
       "  'Ciliated 2',\n",
       "  'Club',\n",
       "  'Dendritic cells',\n",
       "  'Donor',\n",
       "  'Endothelial',\n",
       "  'Fibroblasts',\n",
       "  'Luminal_Macrophages',\n",
       "  'Lymphatic',\n",
       "  'Mast cells',\n",
       "  'Mesothelium',\n",
       "  'Neutrophils',\n",
       "  'Sample',\n",
       "  'Secretory',\n",
       "  'Smooth muscle',\n",
       "  'T and NK',\n",
       "  'Type 2 alveolar',\n",
       "  'Type_1_alveolar'},\n",
       " 'https://cellxgene.cziscience.com/d/cellular_census_of_human_lungs_bronchi-17.cxg/': {'A cellular census of human lungs identifies novel cell states in health and in asthma - bronchi',\n",
       "  'B cells',\n",
       "  'Basal',\n",
       "  'Basal 1',\n",
       "  'Basal 2',\n",
       "  'BroadCellType',\n",
       "  'CellType',\n",
       "  'Ciliated',\n",
       "  'Ciliated 1',\n",
       "  'Ciliated 2',\n",
       "  'Club',\n",
       "  'Dendritic cells',\n",
       "  'Donor',\n",
       "  'Endothelial',\n",
       "  'Fibroblasts',\n",
       "  'Goblet 2',\n",
       "  'Goblet_1',\n",
       "  'Ionocytes',\n",
       "  'Luminal_Macrophages',\n",
       "  'Lymphatic',\n",
       "  'Mast cells',\n",
       "  'Neutrophils',\n",
       "  'Sample',\n",
       "  'Secretory',\n",
       "  'Smooth muscle',\n",
       "  'Source',\n",
       "  'T and NK'},\n",
       " 'https://cellxgene.cziscience.com/d/cellular_census_of_human_lungs_nasal-16.cxg/': {'A cellular census of human lungs identifies novel cell states in health and in asthma - nasal',\n",
       "  'B cells',\n",
       "  'Basal',\n",
       "  'Basal 1',\n",
       "  'Basal 2',\n",
       "  'BroadCellType',\n",
       "  'CellType',\n",
       "  'Ciliated',\n",
       "  'Ciliated 1',\n",
       "  'Ciliated 2',\n",
       "  'Club',\n",
       "  'Dendritic cells',\n",
       "  'Donor',\n",
       "  'Goblet 2',\n",
       "  'Goblet_1',\n",
       "  'Ionocytes',\n",
       "  'Luminal_Macrophages',\n",
       "  'Neutrophils',\n",
       "  'Sample',\n",
       "  'Secretory'},\n",
       " 'https://cellxgene.cziscience.com/d/Single_cell_gene_expression_profiling_of_SARS_CoV_2_infected_human_cell_lines_H1299-27.cxg/': {'Single-cell gene expression profiling of SARS-CoV-2 infected human cell lines - H1299',\n",
       "  'infect',\n",
       "  'louvain',\n",
       "  'orig_ident',\n",
       "  'sample_id',\n",
       "  'sample_name',\n",
       "  'strain'},\n",
       " 'https://cellxgene.cziscience.com/d/Single_cell_gene_expression_profiling_of_SARS_CoV_2_infected_human_cell_lines_Calu_3-28.cxg/': {'Single-cell gene expression profiling of SARS-CoV-2 infected human cell lines - Calu-3',\n",
       "  'infect',\n",
       "  'louvain',\n",
       "  'orig_ident',\n",
       "  'sample_id',\n",
       "  'sample_name',\n",
       "  'strain'},\n",
       " 'https://cellxgene.cziscience.com/d/ischaemic_sensitivity_of_human_tissue_by_single_cell_RNA_seq_lung-18.cxg/': {'CellType',\n",
       "  'Donor',\n",
       "  'Ischaemic sensitivity of human tissue by single cell RNA seq - lung',\n",
       "  'Time',\n",
       "  'donor_time',\n",
       "  'leiden',\n",
       "  'patient',\n",
       "  'sample'},\n",
       " 'https://cellxgene.cziscience.com/d/gut_cell_atlas-24.cxg/': {'BCR_ChainCombination',\n",
       "  'BCR_ISOTYPE',\n",
       "  'Gut Cell Atlas',\n",
       "  'IGH_J_CALL',\n",
       "  'IGH_V_CALL_GENOTYPED',\n",
       "  'IGL_C_Gene',\n",
       "  'IGL_VDJ_Gene',\n",
       "  'TCR_c_gene',\n",
       "  'TCR_d_gene',\n",
       "  'TCR_j_gene',\n",
       "  'TCR_v_gene',\n",
       "  'cell_type',\n",
       "  'donor',\n",
       "  'region'},\n",
       " 'https://cellxgene.cziscience.com/d/human_fetal_liver_single_cell_transcriptome-13.cxg/': {'Annot',\n",
       "  'BroadCellType',\n",
       "  'Human fetal liver single cell transcriptome data',\n",
       "  'Source',\n",
       "  'batch'},\n",
       " 'https://cellxgene.cziscience.com/d/spatio_temporal_immune_zonation_of_the_human_kidney-21.cxg/': {'Experiment',\n",
       "  'Project',\n",
       "  'Spatio-temporal immune zonation of the human kidney',\n",
       "  'broad_celltype',\n",
       "  'celltype',\n",
       "  'compartment'}}"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'https://cellxgene.cziscience.com/d/cellular_census_of_human_lungs_alveoli_and_parenchyma-15.cxg/': {'A cellular census of human lungs identifies novel cell states in health and in asthma - parenchyma',\n",
       "  'B cells',\n",
       "  'Basal',\n",
       "  'Basal 1',\n",
       "  'Basal 2',\n",
       "  'BroadCellType',\n",
       "  'CellType',\n",
       "  'Ciliated',\n",
       "  'Ciliated 1',\n",
       "  'Ciliated 2',\n",
       "  'Club',\n",
       "  'Dendritic cells',\n",
       "  'Donor',\n",
       "  'Endothelial',\n",
       "  'Fibroblasts',\n",
       "  'Luminal_Macrophages',\n",
       "  'Lymphatic',\n",
       "  'Mast cells',\n",
       "  'Mesothelium',\n",
       "  'Neutrophils',\n",
       "  'Sample',\n",
       "  'Secretory',\n",
       "  'Smooth muscle',\n",
       "  'T and NK',\n",
       "  'Type 2 alveolar',\n",
       "  'Type_1_alveolar'},\n",
       " 'https://cellxgene.cziscience.com/d/cellular_census_of_human_lungs_bronchi-17.cxg/': {'A cellular census of human lungs identifies novel cell states in health and in asthma - bronchi',\n",
       "  'B cells',\n",
       "  'Basal',\n",
       "  'Basal 1',\n",
       "  'Basal 2',\n",
       "  'BroadCellType',\n",
       "  'CellType',\n",
       "  'Ciliated',\n",
       "  'Ciliated 1',\n",
       "  'Ciliated 2',\n",
       "  'Club',\n",
       "  'Dendritic cells',\n",
       "  'Donor',\n",
       "  'Endothelial',\n",
       "  'Fibroblasts',\n",
       "  'Goblet 2',\n",
       "  'Goblet_1',\n",
       "  'Ionocytes',\n",
       "  'Luminal_Macrophages',\n",
       "  'Lymphatic',\n",
       "  'Mast cells',\n",
       "  'Neutrophils',\n",
       "  'Sample',\n",
       "  'Secretory',\n",
       "  'Smooth muscle',\n",
       "  'Source',\n",
       "  'T and NK'},\n",
       " 'https://cellxgene.cziscience.com/d/cellular_census_of_human_lungs_nasal-16.cxg/': {'A cellular census of human lungs identifies novel cell states in health and in asthma - nasal',\n",
       "  'B cells',\n",
       "  'Basal',\n",
       "  'Basal 1',\n",
       "  'Basal 2',\n",
       "  'BroadCellType',\n",
       "  'CellType',\n",
       "  'Ciliated',\n",
       "  'Ciliated 1',\n",
       "  'Ciliated 2',\n",
       "  'Club',\n",
       "  'Dendritic cells',\n",
       "  'Donor',\n",
       "  'Goblet 2',\n",
       "  'Goblet_1',\n",
       "  'Ionocytes',\n",
       "  'Luminal_Macrophages',\n",
       "  'Neutrophils',\n",
       "  'Sample',\n",
       "  'Secretory'},\n",
       " 'https://cellxgene.cziscience.com/d/Single_cell_gene_expression_profiling_of_SARS_CoV_2_infected_human_cell_lines_H1299-27.cxg/': {'Single-cell gene expression profiling of SARS-CoV-2 infected human cell lines - H1299',\n",
       "  'infect',\n",
       "  'louvain',\n",
       "  'orig_ident',\n",
       "  'sample_id',\n",
       "  'sample_name',\n",
       "  'strain'},\n",
       " 'https://cellxgene.cziscience.com/d/Single_cell_gene_expression_profiling_of_SARS_CoV_2_infected_human_cell_lines_Calu_3-28.cxg/': {'Single-cell gene expression profiling of SARS-CoV-2 infected human cell lines - Calu-3',\n",
       "  'infect',\n",
       "  'louvain',\n",
       "  'orig_ident',\n",
       "  'sample_id',\n",
       "  'sample_name',\n",
       "  'strain'},\n",
       " 'https://cellxgene.cziscience.com/d/ischaemic_sensitivity_of_human_tissue_by_single_cell_RNA_seq_lung-18.cxg/': {'Alveolar_Type1',\n",
       "  'Alveolar_Type2',\n",
       "  'B_cells',\n",
       "  'Blood_vessel',\n",
       "  'CellType',\n",
       "  'Ciliated',\n",
       "  'DC_1',\n",
       "  'DC_2',\n",
       "  'DC_Monocyte_Dividing',\n",
       "  'DC_activated',\n",
       "  'DC_plasmacytoid',\n",
       "  'Donor',\n",
       "  'Fibroblast',\n",
       "  'Ischaemic sensitivity of human tissue by single cell RNA seq - lung',\n",
       "  'Lymph_vessel',\n",
       "  'Macrophage_Dividing',\n",
       "  'Macrophage_MARCOneg',\n",
       "  'Macrophage_MARCOpos',\n",
       "  'Mast_cells',\n",
       "  'Monocyte',\n",
       "  'Muscle_cells',\n",
       "  'NK',\n",
       "  'NK_Dividing',\n",
       "  'Plasma_cells',\n",
       "  'T_CD4',\n",
       "  'T_CD8_CytT',\n",
       "  'T_cells_Dividing',\n",
       "  'T_regulatory',\n",
       "  'Time',\n",
       "  'donor_time',\n",
       "  'leiden',\n",
       "  'patient',\n",
       "  'sample'},\n",
       " 'https://cellxgene.cziscience.com/d/gut_cell_atlas-24.cxg/': {'Activated CD4 T',\n",
       "  'B cell IgA Plasma',\n",
       "  'B cell IgG Plasma',\n",
       "  'B cell cycling',\n",
       "  'B cell memory',\n",
       "  'BCR_ChainCombination',\n",
       "  'BCR_ISOTYPE',\n",
       "  'CD8 T',\n",
       "  'Follicular B cell',\n",
       "  'Gut Cell Atlas',\n",
       "  'IGH_J_CALL',\n",
       "  'IGH_V_CALL_GENOTYPED',\n",
       "  'IGL_C_Gene',\n",
       "  'IGL_VDJ_Gene',\n",
       "  'ILC',\n",
       "  'LYVE1 Macrophage',\n",
       "  'Lymphoid DC',\n",
       "  'Macrophage',\n",
       "  'Mast',\n",
       "  'Monocyte',\n",
       "  'NK',\n",
       "  'TCR_c_gene',\n",
       "  'TCR_d_gene',\n",
       "  'TCR_j_gene',\n",
       "  'TCR_v_gene',\n",
       "  'Tcm',\n",
       "  'Tfh',\n",
       "  'Th1',\n",
       "  'Th17',\n",
       "  'Treg',\n",
       "  'cDC1',\n",
       "  'cDC2',\n",
       "  'cell_type',\n",
       "  'cycling DCs',\n",
       "  'cycling gd T',\n",
       "  'donor',\n",
       "  'gd T',\n",
       "  'pDC',\n",
       "  'region'},\n",
       " 'https://cellxgene.cziscience.com/d/human_fetal_liver_single_cell_transcriptome-13.cxg/': {'Annot',\n",
       "  'BroadCellType',\n",
       "  'Endothelial',\n",
       "  'Erythroid',\n",
       "  'Fibroblast',\n",
       "  'Hepatocyte',\n",
       "  'Human fetal liver single cell transcriptome data',\n",
       "  'Source',\n",
       "  'batch',\n",
       "  'immune cells'},\n",
       " 'https://cellxgene.cziscience.com/d/spatio_temporal_immune_zonation_of_the_human_kidney-21.cxg/': {'Ascending vasa recta endothelium',\n",
       "  'B cell',\n",
       "  'CD4 T cell',\n",
       "  'CD8 T cell',\n",
       "  'Connecting tubule',\n",
       "  'Descending vasa recta endothelium',\n",
       "  'Distinct proximal tubule 1',\n",
       "  'Distinct proximal tubule 2',\n",
       "  'Epithelial progenitor cell',\n",
       "  'Experiment',\n",
       "  'Fibroblast',\n",
       "  'Glomerular endothelium',\n",
       "  'Indistinct intercalated cell',\n",
       "  'Intercalated cell',\n",
       "  'MNP-a/classical monocyte derived',\n",
       "  'MNP-b/non-classical monocyte derived',\n",
       "  'MNP-c/dendritic cell',\n",
       "  'MNP-d/Tissue macrophage',\n",
       "  'Mast cell',\n",
       "  'Myofibroblast',\n",
       "  'NK cell',\n",
       "  'NKT cell',\n",
       "  'Neutrophil',\n",
       "  'Pelvic epithelium',\n",
       "  'Peritubular capillary endothelium',\n",
       "  'Peritubular capillary endothelium 1',\n",
       "  'Peritubular capillary endothelium 2',\n",
       "  'Plasmacytoid dendritic cell',\n",
       "  'Podocyte',\n",
       "  'Principal cell',\n",
       "  'Project',\n",
       "  'Proliferating Proximal Tubule',\n",
       "  'Proximal tubule',\n",
       "  'Spatio-temporal immune zonation of the human kidney',\n",
       "  'Thick ascending limb of Loop of Henle',\n",
       "  'Transitional urothelium',\n",
       "  'Type A intercalated cell',\n",
       "  'Type B intercalated cell',\n",
       "  'broad_celltype',\n",
       "  'celltype',\n",
       "  'compartment'},\n",
       " 'https://cellxgene.cziscience.com/d/fetal_maternal_interface_10x-22.cxg/': {'CellType',\n",
       "  'DC1',\n",
       "  'DC2',\n",
       "  'EVT',\n",
       "  'Endo (f)',\n",
       "  'Endo (m)',\n",
       "  'Endo L',\n",
       "  'Granulocytes',\n",
       "  'HB',\n",
       "  'ILC3',\n",
       "  'Location',\n",
       "  'MAIT',\n",
       "  'MO',\n",
       "  'NK CD16+',\n",
       "  'NK CD16-',\n",
       "  'PB Naive CD4 ',\n",
       "  'PB Naive CD8',\n",
       "  'PB clonal CD8',\n",
       "  'Plasma',\n",
       "  'Reconstructing the human first trimester fetal-maternal interface using single cell transcriptomics - 10x',\n",
       "  'SCT',\n",
       "  'Stage',\n",
       "  'Treg',\n",
       "  'VCT',\n",
       "  'dM1',\n",
       "  'dM2',\n",
       "  'dM3',\n",
       "  'dNK p',\n",
       "  'dNK1',\n",
       "  'dNK2',\n",
       "  'dNK3',\n",
       "  'dP1',\n",
       "  'dP2',\n",
       "  'dS1',\n",
       "  'dS2',\n",
       "  'dS3',\n",
       "  'dT CD4',\n",
       "  'dT CD8',\n",
       "  'fFB1',\n",
       "  'fFB2'},\n",
       " 'https://cellxgene.cziscience.com/d/fetal_maternal_interface_smartseq2-23.cxg/': {'CellType',\n",
       "  'DC1',\n",
       "  'DC2',\n",
       "  'EVT',\n",
       "  'Endo (m)',\n",
       "  'Endo L',\n",
       "  'Granulocytes',\n",
       "  'HB',\n",
       "  'ILC3',\n",
       "  'Location',\n",
       "  'MAIT',\n",
       "  'MO',\n",
       "  'NK CD16+',\n",
       "  'NK CD16-',\n",
       "  'PB Naive CD4 ',\n",
       "  'PB Naive CD8',\n",
       "  'PB clonal CD8',\n",
       "  'Plasma',\n",
       "  'Reconstructing the human first trimester fetal-maternal interface using single cell transcriptomics - SmartSeq2',\n",
       "  'SCT',\n",
       "  'Treg',\n",
       "  'VCT',\n",
       "  'dM1',\n",
       "  'dM2',\n",
       "  'dM3',\n",
       "  'dNK p',\n",
       "  'dNK1',\n",
       "  'dNK2',\n",
       "  'dNK3',\n",
       "  'dP1',\n",
       "  'dP2',\n",
       "  'dS1',\n",
       "  'dS2',\n",
       "  'dS3',\n",
       "  'dT CD4',\n",
       "  'dT CD8',\n",
       "  'fFB1',\n",
       "  'fFB2'},\n",
       " 'https://cellxgene.cziscience.com/d/cell_atlas_of_thymic_development-14.cxg/': {'a',\n",
       "  'd',\n",
       "  'e',\n",
       "  'f',\n",
       "  'i',\n",
       "  'l'},\n",
       " 'https://cellxgene.cziscience.com/d/ischaemic_sensitivity_of_human_tissue_by_single_cell_RNA_seq_spleen-19.cxg/': {'Celltypes',\n",
       "  'Donor',\n",
       "  'Ischaemic sensitivity of human tissue by single cell RNA seq - spleen',\n",
       "  'Time',\n",
       "  'donor_time',\n",
       "  'leiden',\n",
       "  'patient',\n",
       "  'sample'},\n",
       " 'https://cellxgene.cziscience.com/d/ischaemic_sensitivity_of_human_tissue_by_single_cell_RNA_seq_oesophagus-20.cxg/': {'Celltypes',\n",
       "  'Donor',\n",
       "  'Ischaemic sensitivity of human tissue by single cell RNA seq - oesophagus',\n",
       "  'Time',\n",
       "  'donor_time',\n",
       "  'leiden',\n",
       "  'patient',\n",
       "  'sample'}}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['https://cellxgene.cziscience.com/d/cellular_census_of_human_lungs_alveoli_and_parenchyma-15.cxg/',\n",
       " 'https://cellxgene.cziscience.com/d/cellular_census_of_human_lungs_bronchi-17.cxg/',\n",
       " 'https://cellxgene.cziscience.com/d/cellular_census_of_human_lungs_nasal-16.cxg/',\n",
       " 'https://cellxgene.cziscience.com/d/Single_cell_gene_expression_profiling_of_SARS_CoV_2_infected_human_cell_lines_H1299-27.cxg/',\n",
       " 'https://cellxgene.cziscience.com/d/Single_cell_gene_expression_profiling_of_SARS_CoV_2_infected_human_cell_lines_Calu_3-28.cxg/',\n",
       " 'https://cellxgene.cziscience.com/d/ischaemic_sensitivity_of_human_tissue_by_single_cell_RNA_seq_lung-18.cxg/',\n",
       " 'https://cellxgene.cziscience.com/d/gut_cell_atlas-24.cxg/',\n",
       " 'https://cellxgene.cziscience.com/d/human_fetal_liver_single_cell_transcriptome-13.cxg/',\n",
       " 'https://cellxgene.cziscience.com/d/spatio_temporal_immune_zonation_of_the_human_kidney-21.cxg/',\n",
       " 'https://cellxgene.cziscience.com/d/fetal_maternal_interface_10x-22.cxg/',\n",
       " 'https://cellxgene.cziscience.com/d/fetal_maternal_interface_smartseq2-23.cxg/',\n",
       " 'https://cellxgene.cziscience.com/d/cell_atlas_of_thymic_development-14.cxg/',\n",
       " 'https://cellxgene.cziscience.com/d/ischaemic_sensitivity_of_human_tissue_by_single_cell_RNA_seq_spleen-19.cxg/',\n",
       " 'https://cellxgene.cziscience.com/d/ischaemic_sensitivity_of_human_tissue_by_single_cell_RNA_seq_oesophagus-20.cxg/']"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
